components/match2/wikis/stormgate/match_group_util_custom.lua

```diff
@@ -0,0 +1,373 @@
+---
+-- @Liquipedia
+-- wiki=stormgate
+-- page=Module:MatchGroup/Util/Custom
+--
+-- Please see https://github.com/Liquipedia/Lua-Modules to contribute
+--
+
+local Array = require('Module:Array')
+local Faction = require('Module:Faction')
+local Flags = require('Module:Flags')
+local Logic = require('Module:Logic')
+local Lua = require('Module:Lua')
+local String = require('Module:StringUtils')
+local Table = require('Module:Table')
+local TypeUtil = require('Module:TypeUtil')
+
+local MatchGroupUtil = Lua.import('Module:MatchGroup/Util')
+-- can not use `Module:OpponentLibraries`/`Module:Opponent/Custom` to avoid loop
+local Opponent = Lua.import('Module:Opponent')
+
+local SCORE_STATUS = 'S'
+
+local CustomMatchGroupUtil = Table.deepCopy(MatchGroupUtil)
+
+CustomMatchGroupUtil.types.Faction = TypeUtil.literalUnion(unpack(Faction.getFactions()))
+
+CustomMatchGroupUtil.types.Player = TypeUtil.extendStruct(MatchGroupUtil.types.Player, {
+	position = 'number?',
+	faction = CustomMatchGroupUtil.types.Faction,
+	random = 'boolean',
+})
+
+---@class StormgateMatchGroupUtilGamePlayer: StormgateStandardPlayer
+---@field matchplayerIndex integer
+---@field heroes string[]?
+---@field position integer
+---@field random boolean
+
+---@class StormgateMatchGroupUtilGameOpponent:GameOpponent
+---@field placement number?
+---@field players StormgateMatchGroupUtilGamePlayer[]
+---@field score number?
+CustomMatchGroupUtil.types.GameOpponent = TypeUtil.struct({
+	placement = 'number?',
+	players = TypeUtil.array(CustomMatchGroupUtil.types.Player),
+	score = 'number?',
+})
+
+---@class StormgateMatchGroupUtilGame: MatchGroupUtilGame
+---@field opponents  StormgateMatchGroupUtilGameOpponent[]
+---@field offFactions table<integer, string[]>?
+
+---@class StormgateMatchGroupUtilVeto
+---@field by number?
+---@field map string
+
+---@class StormgateMatchGroupUtilSubmatch
+---@field games StormgateMatchGroupUtilGame[]
+---@field mode string
+---@field opponents StormgateMatchGroupUtilGameOpponent[]
+---@field resultType ResultType
+---@field scores table<number, number>
+---@field subgroup number
+---@field walkover WalkoverType
+---@field winner number?
+---@field header string?
+
+---@class StormgateMatchGroupUtilMatch: MatchGroupUtilMatch
+---@field games StormgateMatchGroupUtilGame[]
+---@field isFfa boolean
+---@field noScore boolean?
+---@field opponents StormgateStandardOpponent[]
+---@field vetoes StormgateMatchGroupUtilVeto[]
+---@field submatches StormgateMatchGroupUtilSubmatch[]?
+---@field casters string?
+---@field isUniformMode boolean
+
+---@param record table
+---@return StormgateMatchGroupUtilMatch
+function CustomMatchGroupUtil.matchFromRecord(record)
+	local match = MatchGroupUtil.matchFromRecord(record) --[[@as StormgateMatchGroupUtilMatch]]
```

Comment:
Not on a device with LuaLS so can't test, but does this work?
If so, IMO preferable, at least for the local variable definitions.
```suggestion
    ---@type StormgateMatchGroupUtilMatch
	local match = MatchGroupUtil.matchFromRecord(record)
```

Reply:
![Screenshot 2024-01-29 163110](https://github.com/Liquipedia/Lua-Modules/assets/75081997/980052ad-f6b6-4246-8fef-ac69f96dc4a2)